### ISSUE TITLE
Garbage collection in Fysom

### ISSUE THREAD

**Issue by sjlongland** [mriehl/fysom#21]
Hi,

I recently stumbled across this project and it looks to be a very well contained library.  One concern I have is the use of callbacks with garbage collection.

So I have an object with a state machine, and a few callbacks:
```
import fysom
import gc
class MyObject(object):
    def __init__(self, *args, **kwargs):
        self.state_machine = fysom.Fysom(
            initial='a',
            events=[
                ('go_to_b', 'a', 'b'),
                ('go_to_a', 'b', 'a'),
            ],
            callbacks={
                'ona': self._on_a,
                'onb': self._on_b,
            }
        )
    def __del__(self):
        print ('Goodbye from %s' % self)
    def _on_a(self, *args, **kwargs):
        print ('In state A: args=%s, kwargs=%s' % (args, kwargs))
    def _on_b(self, *args, **kwargs):
        print ('In state B: args=%s, kwargs=%s' % (args, kwargs))

obj = MyObject()
obj.state_machine.go_to_b(123)
obj.state_machine.go_to_a(234)
del obj

# This should be empty, but isn't due to circular reference
print (filter(lambda o : isinstance(o, MyObject), gc.get_objects()))
```

If I run that, I get this:
```
RC=0 stuartl@rikishi /tmp $ python test.py 
In state A: args=(<fysom._e_obj object at 0x7f3fe6898b50>,), kwargs={}
In state B: args=(<fysom._e_obj object at 0x7f3fe6898b50>,), kwargs={}
In state A: args=(<fysom._e_obj object at 0x7f3fe6898b50>,), kwargs={}
[<__main__.MyObject object at 0x7f3fe6886d10>]
```

Is there some way that I can structure my code to prevent this from happening?

**Comment by mriehl** [comment]
Very nice catch, this seems like it would cause some really unexpected annoyances (I think we have an app running in production, where this problem could exist (though we never experienced problems yet).

I'll review, merge and release your changes ASAP.

**Comment by mriehl** [comment]
Fixed with `2.1.0` thanks to your pull request.
I'm not able to GPG sign the release right now but will do so once I get the chance.

**Comment by sjlongland** [comment]
Thanks for merging this.  One thing I note is that I really should have done some looking around before implementing my own weakref-based method referencing:

https://github.com/twang817/weakrefmethod

Seems there's a module to do it for Python 2.x.  I didn't realise it was standard in 3.4 and I note this module misses out on 3.0-3.3 (which my patch seems to take care of).  I might look at getting that module tweaked to handle 3.0-3-3 (and become a no-op on ≥3.4).
